File: snowy_daemon/src/handler/queue.rs
use snowy_core::library::TrackId;

use crate::states::AppState;

pub async fn enqueue(id: TrackId, state: &mut AppState) {
    if let Some(track) = state.library.tracks.get(&id) {
        state.orchestrator.queue.enqueue(track.clone())
    }
}

pub async fn prepend(id: TrackId, state: &mut AppState) {
    if let Some(track) = state.library.tracks.get(&id) {
        state.orchestrator.queue.prepend_queue(track.clone())
    }
}

pub async fn dequeue(index: usize, state: &mut AppState) {
    state.orchestrator.queue.dequeue(index);
}

pub async fn next(state: &mut AppState) {
    state.orchestrator.queue.next();
}

pub async fn prev(state: &mut AppState) {
    state.orchestrator.queue.prev();
}

pub async fn shuffle(state: &mut AppState) {
    state.orchestrator.queue.shuffle();
}

pub async fn clear(state: &mut AppState) {
    state.orchestrator.queue.clear();
}

File: web-ui/src/lib/commands/queue.ts
import type { LoopMode } from '../../types/loop_mode';
import type { TrackId } from '../../types/track_id';
import { send } from '../ws';

export function enqueue(track_id: string) {
	send({ Queue: { Enqueue: { track_id } } });
}

export function prepend(track_id: TrackId) {
	send({ Queue: { Prepend: { track_id } } });
}

export function dequeue(index: number) {
	send({ Queue: { Dequeue: { index } } });
}

export function playNow(track_id: TrackId) {
	send({ Queue: { PlayNow: { track_id } } });
}

export function next() {
	send({ Queue: 'Next' });
}

export function prev() {
	send({ Queue: 'Prev' });
}

export function shuffle() {
	send({ Queue: 'Shuffle' });
}

export function clear() {
	send({ Queue: 'Clear' });
}

export function setLoopMode(mode: LoopMode) {
	send({ Queue: { SetLoopMode: mode } });
}

export function cycleLoopMode() {
	send({ Queue: 'CycleLoopMode' });
}

export function fetch() {
	send({ Queue: 'Fetch' });
}
